src/mongo/db/pipeline/expression.cpp

```diff
@@ -2716,31 +2730,102 @@ Value ExpressionIndexOfArray::evaluate(const Document& root) const {
         // Don't let 'endIndex' exceed the length of the array.
-        endIndex = std::min(array.size(), static_cast<size_t>(endIndexArg.coerceToInt()));
+        endIndex = std::min(arrayLength, static_cast<size_t>(endIndexArg.coerceToInt()));
+    }
+    deps.push_back(Value(endIndex));
+    return deps;
+}
+/**
+ *  This class handles the case where IndexOfArray is given an ExpressionConstant
+ *  instead of using a vector and searching through it we can use a unordered_map
+ *  for O(1) lookup time.
+ **/
+class ExpressionIndexOfArray::Optimized : public ExpressionIndexOfArray {
+public:
+    Optimized(const boost::intrusive_ptr<ExpressionContext>& expCtx,
+              const ValueUnorderedMap<int>& indexMap,
+              const ExpressionVector& operands)
+        : ExpressionIndexOfArray(expCtx), _indexMap(std::move(indexMap)) {
+        vpOperand = operands;
     }
 
-    for (size_t i = startIndex; i < endIndex; i++) {
-        if (getExpressionContext()->getValueComparator().evaluate(array[i] == searchItem)) {
-            return Value(static_cast<int>(i));
+    virtual Value evaluate(const Document& root) const {
+        std::vector<Value> operands = parseDeps(root, vpOperand, _indexMap.size());
+        auto index = _indexMap.find(operands[0]);
+        auto comparator = getExpressionContext()->getValueComparator();
+        Value vIndex = Value(index->second);
+        if (index != _indexMap.end() && comparator.evaluate(Value(index->second) >= operands[1]) &&
+            comparator.evaluate(Value(index->second) < operands[2])) {
+            return Value(index->second);
+        } else {
+            // If the item we are searching is not found or if not found in given range return -1.
+            return Value(-1);
         }
     }
 
-    return Value(-1);
+private:
+    const ValueUnorderedMap<int> _indexMap;
+};
+
+intrusive_ptr<Expression> ExpressionIndexOfArray::optimize() {
+    ExpressionNary::optimize();
+
+    // If the input arr is an ExpressionConstant we can optimize using a unordered_map instead of an
+    // Array
+    if (dynamic_cast<ExpressionConstant*>(vpOperand[0].get())) {
+        ExpressionConstant* ec = dynamic_cast<ExpressionConstant*>(vpOperand[0].get());
+        const Value valueArray = ec->getValue();
+        if (valueArray.nullish()) {
+            return this;
+        }
+        uassert(50749,
+                str::stream() << "First operand of $indexOfArray must be an array. First "
+                              << "argument is of type: "
+                               << typeName(valueArray.getType()),
+                valueArray.isArray());
+
+        auto arr = valueArray.getArray();
+
+        ValueUnorderedMap<int> indexMap =
+            getExpressionContext()->getValueComparator().makeUnorderedValueMap<int>();
+
+        for (int i = 0; i < int(arr.size()); i++) {
+            auto pair = std::make_pair(arr[i], i);
```

Comment:
optional, re-write this as
`indexMap.emplace(arr[i], i);` or if that doesn't work, `indexMap[arr[i]] = i;`? I'm curious why you made a pair here.